src/lib/optimizer/strategy/data_induced_predicate_rule.cpp

```diff
@@ -0,0 +1,130 @@
+            expression_vector(), expression_vector(min_(reducer_side_expression), max_(reducer_side_expression)),
+            reducer_node);
+
+        const auto min = ProjectionNode::make(expression_vector(min_(reducer_side_expression)), subquery);
+        const auto max = ProjectionNode::make(expression_vector(max_(reducer_side_expression)), subquery);
+
+        // The between predicate is built using the projected min and max from the reducer side expression.
+        auto between_predicate =
+            PredicateNode::make(between_inclusive_(reduced_side_expression, lqp_subquery_(min), lqp_subquery_(max)));
+        lqp_insert_node(join_node, selection_side, between_predicate);
+
+        const auto reduced_cardinality = estimator->estimate_cardinality(between_predicate);
+        lqp_remove_node(between_predicate);
+
+        // If the `MINIMUM_SELECTIVITY` is not met we will not consider this join_predicate for this side.
+        if (original_cardinality == 0 || (reduced_cardinality / original_cardinality) > MINIMUM_SELECTIVITY) {
+          return;
+        }
+        data_induced_predicates.emplace_back(join_node, selection_side, between_predicate);
+        /**
+         * Additionally to this, future work could focus on evaluating if a heuristic that inserts semi join reductions
+         * if no predicate is found could be of further interest. This could be done by searching for a predicate in the
+         * lqp and adding a semi join reduction if the minimum selectivity of the SJR rule is met. First attempts for
+         * this are in the commit history (commit f6c5fb5) of this PR.
+         *
+         * Searching for a predicate prior to the join must not ensure a good DIP performance. In contrast to that the
+         * source paper mentioned in this PR suggests Range Sets as a solution. This could improve the performance
+         * drastically.
+         */
+      };
+
+      // Having defined the lambda responsible for conditionally adding a data induced predicate, we now apply it to
+      // both inputs of the join. For outer joins, we must not filter the side on which tuples survive even without a
+      // join partner.
+      if (join_node->join_mode != JoinMode::FullOuter) {
+        reduce_if_beneficial(LQPInputSide::Right);
+      }
+
+      // On the left side we must not create data induced predicates for anti joins as those rely on the very existence
+      // of non-matching values on the right side.
+      if (join_node->join_mode != JoinMode::FullOuter && join_node->join_mode != JoinMode::AntiNullAsTrue &&
+          join_node->join_mode != JoinMode::AntiNullAsFalse) {
+        reduce_if_beneficial(LQPInputSide::Left);
+      }
+    }
+    return LQPVisitation::VisitInputs;
+  });
+
+  for (const auto& [join_node, side_of_join, data_induced_predicate_node] : data_induced_predicates) {
+    lqp_insert_node(join_node, side_of_join, data_induced_predicate_node, AllowRightInput::Yes);
```

Comment:
I don't think this makes any difference in this case. You are right this could be change, so both lines would be the same.
For the cardinality estimation it is not necessary, because we remove it afterwards.
But for our reduction node in the end we don't want a right input. We just want the input, which will be reduced afterwards.

Reply:
This flag is about if the inserted node (the diP in this case) is allowed to have a right input when inserted into the LQP (LQP insertion means that we we link the node to insert as the left input of the parent node and re-wire the parent's original left input to be the left input of the inserted node). As the diP is a PredicateNode w/o right input, we don't have to allow the inserted node to have a right input set (different from semi-join reduction: there, the reducer has a right input, which is set accordingly and which we want to keep).